lib/worker/output.go

```diff
@@ -0,0 +1,219 @@
+	readers []io.Reader
+	// combinedBuffer keeps the combined output, this is used to forward to
+	// new listeners that came in late
+	combinedBuffer []byte
+	// messages is a channel where new output is delivered
+	messages chan []byte
+	// errors is a channel to forward output reading errors to listeners
+	errors chan error
+	// done is a channel to notify listeners that all output has been read
+	done chan struct{}
+	mu   *sync.Mutex
+}
+
+// ProcessOutputEntry is the struct sent to clients' channels
+// streaming output
+type ProcessOutputEntry struct {
+	Content    []byte
+	ReceivedAt time.Time
+}
+
+// Listener represents an output listener that the output handler will
+// send updates to.
+type Listener struct {
+	outputChan chan ProcessOutputEntry
+	errorChan  chan error
+}
+
+// NewOutputHandler returns a new OutputHandler struct from the provided readers.
+func NewOutputHandler(rc ...io.Reader) (*OutputHandler, error) {
+	if len(rc) == 0 {
+		return nil, fmt.Errorf("must provide at least one io.Reader")
+	}
+	o := &OutputHandler{
+		readers:  rc,
+		mu:       &sync.Mutex{},
+		done:     make(chan struct{}),
+		messages: make(chan []byte),
+		errors:   make(chan error),
+	}
+	go o.handleOutput()
+	go o.handleBroadcast()
+	go o.handleErrors()
+	return o, nil
+}
+
+// Stream returns a channel to receive output up to the
+// present time and stream real time. The channel closes when
+// there the process the output is coming from is finished.
+// Any errors encountered during output reading are sent through
+// the errors channel.
+func (o *OutputHandler) Stream() (chan ProcessOutputEntry, chan error) {
+	output := make(chan ProcessOutputEntry)
+	errChan := make(chan error)
+
+	go func() {
+		if o.bufferLen() > 0 {
+			output <- ProcessOutputEntry{Content: o.combinedBuffer}
+		}
+		select {
+		case <-o.done:
+			close(output)
+		default:
+			o.addListener(&Listener{
+				outputChan: output,
+				errorChan:  errChan,
+			})
+		}
+	}()
+
+	return output, errChan
+}
+
+func (o *OutputHandler) addListener(newListener *Listener) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	o.listeners = append(o.listeners, newListener)
+}
+
+func (o *OutputHandler) updateCombinedBuffer(b []byte) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	o.combinedBuffer = append(o.combinedBuffer, b...)
+}
+
+func (o *OutputHandler) bufferLen() int {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return len(o.combinedBuffer)
+}
+
+func (o *OutputHandler) iterListeners(routine func(*Listener)) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	for _, l := range o.listeners {
+		routine(l)
+	}
+}
+
+// handleBroadcast forwads new messages to the listeners and
+// closes listener channels when done signal is received to
+// signal end of stream.
+func (o *OutputHandler) handleBroadcast() {
+L:
+	for {
+		select {
+		case msg := <-o.messages:
+			o.iterListeners(func(l *Listener) {
+				l.outputChan <- ProcessOutputEntry{Content: msg}
```

Comment:
Can a slow listener block the all the listeners on this process here? Not sure how much time/effort you'd want to spend to mitigate that in an exercise like this, though

Reply:
Yes you are right this could happen. 

There's 2 solutions I can think of:
1. A simple solution could be to make the `outputChan` into a buffered channel but this would only delay the issue instead of solving it completely.

2. Instead of sending output directly, do it  non blocking via a select with a default case that maintains a queue of messages per channel but maybe things start getting a bit complex at this point.

Reply:
For this exercise, I would go for the simplest solution (1).